### PR TITLE
Implemented more granular parameters for Telegram notifications

### PR DESCRIPTION
**Pull request by DBa2016**
## Short Description:
"alert_catch" now can be either a list as before (alert on every catch of the corresponding pokemon) or a dictionary of dictionaries, of the form:
"alert_catch": {
   "PokemonName": { "operator": "and/or", "iv": 0.9, "cp": 1000}
}

### CODE REVIEW

**Comment by Gobberwart** [comment]
You know, I quite like this concept, but its association with Telegram puts me off. I really don't want (and won't use) yet another messaging application.

Can I suggest you think about separating the various information-generating code from the actual message-sending code. That way, the information-generating stuff could be re-used to send notifications via other protocols, ie. email, skype, messenger etc.

Eg.

delivery_system = telegram/skype/whatever (obviously needs supporting code)
event = level_up/catch_pokemon/awarded_badge etc.
msg = nicely formatted notification message

then..

delivery_system.send_message(msg)

Let the delivery_system code worry about how to delivery the message, while the rest just deals with handling events and passing them on.

Just my 2 cents. Hope it makes sense.

**Comment by DBa2016** [comment]
@Gobberwart : the original Telegram integration does not originate from me, I only adjusted the notification config and some other things.

In principle, what you are suggesting makes sense, and it would involve the following:
1) define an interface and create an abstraction layer which the actual "information-generating" code will be using
2) create an implementation for every needed IM system, starting with Telegram (because it is already there and should not be removed without a good reason).

None of these two parts is depending nor anyhow related to this PR. You are welcome to go ahead and implement your wish - I will very unlikely do because I am okay with Telegram and don't see a need for further IMs. :)

**Comment by Gobberwart** [comment]
@DBa2016 No problem, just thought I'd comment while I had the opportunity. Never know who's watching. Not something I'm planning to do anything about immediately, but if anyone wants to run with it, go nuts :)

**Comment by solderzzc** [comment]
@Gobberwart  The master setting is whom you can send for. Please check the code inside and correct me if I'm wrong.

**Comment by Gobberwart** [comment]
@solderzzc The master setting? Sorry, not following you.

**Comment by solderzzc** [comment]
@Gobberwart this one:
  "config": {
             "enabled": false,
             "master": null,

You can set the master to your own id, or the message will not send out.

**Comment by solderzzc** [comment]
@Gobberwart  sure, get it.

File: pokemongo_bot/event_handlers/telegram_handler.py [hunk 1]
```diff
@@ -21,10 +21,22 @@ def handle_event(self, event, sender, level, formatted_msg, data):
...
+                        return
                 else:
-                    return
+                    if data["pokemon"] in self.pokemons:
+                        trigger = self.pokemons[data["pokemon"]]
+                    elif "all" in self.pokemons:
+                        trigger = self.pokemons["all"]
+                    else:
+                        return
+                    if (not "operator" in trigger or trigger["operator"] == "and") and data["cp"] >= trigger["cp"] and data["iv"] >= trigger["iv"] or ("operator" in trigger and trigger["operator"] == "or" and (data["cp"] >= trigger["cp"] or data["iv"] >= trigger["iv"])):
```
Comment (solderzzc) [inline pokemongo_bot/event_handlers/telegram_handler.py]: There's good logic operator code in the pokemon catch worker, would you like to use that code @DBa2016

**Comment by solderzzc** [comment]
`LOGIC_TO_FUNCTION = {
    'or': lambda x, y, z: x or y or z,
    'and': lambda x, y, z: x and y and z,
    'orand': lambda x, y, z: x or y and z,
    'andor': lambda x, y, z: x and y or z
}`

**Comment by Gobberwart** [comment]
Does this work with "any" or "all"?

**Comment by solderzzc** [comment]
Seems the `all` is included.
elif "all" in self.pokemons:
+                        trigger = self.pokemons["all"]
+                    else:

**Comment by Gobberwart** [comment]
Cheers mate, I was too lazy to look :)

**Comment by solderzzc** [comment]
Let's have it merge, leave the code fine tune later.

**Comment by solderzzc** [comment]
👍

**Comment by solderzzc** [comment]
@Gobberwart Do you want to try the pull approval ?

**Comment by solderzzc** [comment]
yes, I think so...

**Comment by Gobberwart** [comment]
Yay! :)

**Pull request merged**